.github/workflows/ci.yml

```diff
@@ -122,8 +149,30 @@ jobs:
       - name: Run type checking
         run: uv run tox -e typecheck
 
-  run-workspace-tests:
-    name: Run workspace tests
+  quality-checks-rust:
+    name: Global Quality Checks (Rust)
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request' || (github.event_name == 'push' && github.ref == 'refs/heads/main')
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Install Rust toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          toolchain: stable
+          components: rustfmt, clippy
+
+      - name: Check formatting for all Rust packages
+        run: |
+          find packages -name Cargo.toml -print0 | xargs -0 -I {} bash -c 'cd "$(dirname {})" && cargo fmt --all -- --check'
+
+      - name: Linting with Clippy for all Rust packages
+        run: |
+          find packages -name Cargo.toml -print0 | xargs -0 -I {} bash -c 'cd "$(dirname {})" && cargo clippy --all-targets -- -D warnings'
```

Comment:
https://github.com/rust-lang/miri too? (never used it)

Reply:
I don't think this is really necessary and also requires a nightly rust toolchain